### DISCUSSION TITLE
Element::setFieldValue() could compare values before marking customfield dirty

### DISCUSSION THREAD

**Discussion by leevigraham** [craftcms/cms#8859] [Ideas]
### Description

Still working through this one a bit but… I wonder if Element::setFieldValue() should compare values (strict) and not set the value if they are equal. This would reduce the number of dirty values and attributes.

I think there might be some overhead here if the value is currently not set which might lead to extra db calls.

**Comment by brandonkelly** [comment]
As of Craft 3.4, only changed values are submitted in the POST data for built-in element types’ edit pages (as well as any third party element types that support [delta field registration](https://docs.craftcms.com/v3/extend/element-types.html#saving-field-content-deltas). So in theory `setFieldValue()` will only be called for actually dirty values, most of the time.

We took that approach instead of a back-end solution because there’s no performant way to determine whether fields have changed on the back-end. We’d need to end up normalizing all of the incoming values, and asking the field types to compare the incoming values with the current values, at which point we’re not achieving any significant performance gain.

That said, there is an argument for us being smarter about this from a content governance perspective, so will think about it.

(Semi-related: #5503)

**Comment by leevigraham** [comment]
I'm using a standard checkbox group field and it looks like they are still marked as dirty. I'll double check.
